pkg/checks/traceroute/traceroute.go

```diff
@@ -0,0 +1,227 @@
+	return &Traceroute{
+		config:     Config{},
+		traceroute: newTraceroute,
+		CheckBase: checks.CheckBase{
+			Mu:       sync.Mutex{},
+			DoneChan: make(chan struct{}),
+		},
+	}
+}
+
+type Traceroute struct {
+	checks.CheckBase
+	config     Config
+	traceroute tracerouteFactory
+}
+
+type tracerouteFactory func(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error)
+
+func newTraceroute(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error) {
+	opts := &traceroute.TracerouteOptions{}
+	opts.SetTimeoutMs(timeout)
+	opts.SetRetries(retries)
+	opts.SetMaxHops(maxHops)
+	opts.SetPort(port)
+	return traceroute.Traceroute(dest, opts)
+}
+
+type Result struct {
+	// The minimum amount of hops required to reach the target
+	NumHops int
+	// The path taken to the destination
+	Hops []Hop
+}
+
+type Hop struct {
+	Addr    string
+	Latency time.Duration
+	Success bool
+}
+
+func (d *Traceroute) Run(ctx context.Context, cResult chan checks.ResultDTO) error {
+	ctx, cancel := logger.NewContextWithLogger(ctx)
+	defer cancel()
+	log := logger.FromContext(ctx)
+	log.Info("Starting dns check", "interval", d.config.Interval.String())
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Error("Context canceled", "err", ctx.Err())
+			return ctx.Err()
+		case <-d.DoneChan:
+			return nil
+		case <-time.After(d.config.Interval):
+			res := d.check(ctx)
+
+			cResult <- checks.ResultDTO{
+				Name: d.Name(),
+				Result: &checks.Result{
+					Data:      res,
+					Timestamp: time.Now(),
+				},
+			}
+			log.Debug("Successfully finished dns check run")
+		}
+	}
+}
+
+func (c *Traceroute) GetConfig() checks.Runtime {
+	c.Mu.Lock()
+	defer c.Mu.Unlock()
+	return &c.config
+}
+
+func (c *Traceroute) check(ctx context.Context) map[string]Result {
+	res := make(map[string]Result)
+	log := logger.FromContext(ctx)
+
+	type internalResult struct {
+		addr string
+		res  Result
+	}
+
+	var wg sync.WaitGroup
+	cResult := make(chan internalResult, len(c.config.Targets))
+
+	for _, t := range c.config.Targets {
+		wg.Add(1)
+		go func(t Target) {
+			defer wg.Done()
+			log.Debug("Running traceroute", "target", t.Addr)
+			start := time.Now()
+			tr, trerr := c.traceroute(t.Addr, int(t.Port), int(c.config.Timeout/time.Millisecond), c.config.Retries, c.config.MaxHops)
+			duration := time.Since(start)
+			if trerr != nil {
+				log.Error("Error running traceroute", "err", trerr, "target", t.Addr)
+			}
+
+			log.Debug("Ran traceroute", "result", tr, "duration", duration)
+
+			result := Result{
+				NumHops: len(tr.Hops),
+				Hops:    []Hop{},
+			}
+
+			for _, h := range tr.Hops {
+				result.Hops = append(result.Hops, Hop{
+					Addr:    h.Host,
+					Latency: h.ElapsedTime,
+					Success: h.Success,
+				})
+			}
+			cResult <- internalResult{addr: t.Addr, res: result}
+		}(t)
+	}
+
+	log.Debug("Waiting for traceroute checks to finish")
+
+	go func() {
+		wg.Wait()
+		close(cResult)
+	}()
```

Comment:
Is this go routine necessary? I think we can `defer close(cResult)` after initialization.

Reply:
This is necessary. The loop below reads results from cResult until this channel closed. If we don't close it in some way, the thread would block forever, as no more result are created. By waiting for the waitgroup to finish in a separate goroutine, we can close the channel once we know all targets are done checking. The loop will the run until all results are taken out of the channel and then break